### PR TITLE
feat: update policy definition update

### LINKED ISSUE #2509: Update of `PolicyDefinition`

### PR DESCRIPTION
**Pull request by git-masoud**
## What this PR changes/adds

Adds the UPDATE endpoint to the PolicyDefinitionApi:

- adds one PUT endpoint to the PolicyDefinitionApiController
- adds new update() method to the PolicyDefinitionStore
- adds implementations for CosmosPolicyDefinitionStore, InMemoryPolicyDefinitionStore, SqlPolicyDefinitionStore
- adds tests for all the implementations

## Why it does that

To enable update semantics on PolicyDefinition.

## Further notes

- Created PolicyDefinitionUpdateDto, to exclude the ID. The policyId is passed as path param.
- Created PolicyDefinitionDto as a Parent Dto for PolicyDefinitionUpdateDto and PolicyDefinitionUpdateDto
- Updating Policy will fail if the Policy isn't found.

## Linked Issue(s)

Closes #2509

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [Etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md) for details_)

### CODE REVIEW

File: spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java [hunk 1]
```diff
@@ -64,4 +64,15 @@ public interface PolicyDefinitionService {
...
+    /**
+     * Updates a policy. If the policy does not yet exist, {@link ServiceResult#notFound(String)} will be returned.
+     *
+     * @param policyId the ID of the policy to update
+     * @param policy the contents of the policy.
+     *
+     * @return successful if updated, a failure otherwise.
+     */
+    @NotNull
+    ServiceResult<Void> update(String policyId, PolicyDefinition policy);
```
Comment (jimmarino) [inline spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java:77]: Why is the id explicitly passed here?

File: spi/control-plane/policy-spi/src/main/java/org/eclipse/edc/connector/policy/spi/store/PolicyDefinitionStore.java [hunk 2]
```diff
@@ -55,6 +55,15 @@ public interface PolicyDefinitionStore {
      */
     void save(PolicyDefinition policy);
 
+    /**
+     * Updates the policy.
+     *
+     * @param policyId ID of the Policy to be updated
+     * @param policy to be updated.
+     * @throws EdcPersistenceException if something goes wrong.
```
Comment (jimmarino) [inline spi/control-plane/policy-spi/src/main/java/org/eclipse/edc/connector/policy/spi/store/PolicyDefinitionStore.java:63]: What could go wrong?

File: spi/control-plane/policy-spi/src/testFixtures/java/org/eclipse/edc/connector/policy/spi/testfixtures/store/PolicyDefinitionStoreTestBase.java [hunk 3]
```diff
@@ -90,6 +90,38 @@ void save_alreadyExists() {
...
+
+
+    @Test
+    @DisplayName("Update Policy that does not yet exist")
+    void update_policyDoesNotExist() {
+        var id = getRandomId();
+        var policy = getPolicy(id, "target");
+
+        var updated = getPolicyDefinitionStore().update(id, policy);
+        assertThat(updated).isNull();
```
Comment (jimmarino) [inline spi/control-plane/policy-spi/src/testFixtures/java/org/eclipse/edc/connector/policy/spi/testfixtures/store/PolicyDefinitionStoreTestBase.java:102]: The store should return null in this case. An exception should be raised, or this should be modeled explicitly as a response.

File: resources/openapi/yaml/management-api/policy-definition-api.yaml [hunk 4]
```diff
@@ -207,6 +207,45 @@ paths:
           description: An  policy definition with the given ID does not exist
       tags:
       - Policy
+  /policydefinitions/{policyId}:
+    put:
+      description: "Updates an existing Policy, If the Policy is not found, no further\
```
Comment (jimmarino) [inline resources/openapi/yaml/management-api/policy-definition-api.yaml:212]: An error should be reported

File: extensions/control-plane/store/cosmos/policy-definition-store-cosmos/src/main/java/org/eclipse/edc/connector/store/azure/cosmos/policydefinition/CosmosPolicyDefinitionStore.java [hunk 5]
```diff
@@ -76,6 +78,21 @@ public void save(PolicyDefinition policy) {
         with(retryPolicy).run(() -> cosmosDbApi.saveItem(convertToDocument(policy)));
     }
 
+    @Override
+    public PolicyDefinition update(String policyId, PolicyDefinition policy) {
+        Objects.requireNonNull(policyId, "policyId");
+        Objects.requireNonNull(policy, "policy");
+
+        PolicyDefinition policyDefinition = findById(policyId);
```
Comment (jimmarino) [inline extensions/control-plane/store/cosmos/policy-definition-store-cosmos/src/main/java/org/eclipse/edc/connector/store/azure/cosmos/policydefinition/CosmosPolicyDefinitionStore.java:86]: Please use var

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApi.java [hunk 6]
```diff
@@ -90,4 +91,14 @@ public interface PolicyDefinitionApi {
     )
     void deletePolicy(String id);
 
+    @Operation(description = "Updates an existing Policy, If the Policy is not found, no further action is taken",
```
Comment (jimmarino) [inline extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApi.java:94]: Please see the other comment about the behavior

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java [hunk 7]
```diff
@@ -109,6 +112,23 @@ public IdResponseDto createPolicy(PolicyDefinitionRequestDto requestDto) {
 
     }
 
+    @PUT
+    @Path("{policyId}")
+    @Override
+    public void updatePolicy(@PathParam("policyId") String policyId, PolicyDefinitionUpdateDto updatedPolicyDefinition) {
+        PolicyDefinitionUpdateWrapperDto wrapper = PolicyDefinitionUpdateWrapperDto.Builder.newInstance()
```
Comment (jimmarino) [inline extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java:119]: use var

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/model/PolicyDefinitionDto.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.eclipse.edc.connector.api.management.policy.model;
+
+import jakarta.validation.constraints.NotNull;
+import org.eclipse.edc.policy.model.Policy;
+
+public abstract class PolicyDefinitionDto {
+
+    @NotNull
```
Comment (jimmarino) [inline extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/model/PolicyDefinitionDto.java:22]: This is not needed here

**Review by jimmarino — CHANGES_REQUESTED** [review]

File: spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java (hunk 1 above)
Comment (git-tanuj) [reply spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java:77]: That's correct, we can only use PolicyDefinition.
But, one of the review I saw in #PR [2156](https://github.com/eclipse-edc/Connector/pull/2516/files) states that `ServiceImpl` should not depend upon controllers, so i tried to follow the same.

Please let me know, if I shall change it or maintain the consistent behaviour of `update` semantics in the code base

Thank you.

File: spi/control-plane/policy-spi/src/main/java/org/eclipse/edc/connector/policy/spi/store/PolicyDefinitionStore.java (hunk 2 above)
Comment (git-tanuj) [reply spi/control-plane/policy-spi/src/main/java/org/eclipse/edc/connector/policy/spi/store/PolicyDefinitionStore.java]: Changed the Javadoc
Thank you

File: resources/openapi/yaml/management-api/policy-definition-api.yaml (hunk 4 above)
Comment (git-tanuj) [reply resources/openapi/yaml/management-api/policy-definition-api.yaml]: Sure, I have changed the description

File: extensions/control-plane/store/cosmos/policy-definition-store-cosmos/src/main/java/org/eclipse/edc/connector/store/azure/cosmos/policydefinition/CosmosPolicyDefinitionStore.java (hunk 5 above)
Comment (git-tanuj) [reply extensions/control-plane/store/cosmos/policy-definition-store-cosmos/src/main/java/org/eclipse/edc/connector/store/azure/cosmos/policydefinition/CosmosPolicyDefinitionStore.java]: Sure, made the changes

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApi.java (hunk 6 above)
Comment (git-tanuj) [reply extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApi.java]: Sure, I have changed the description

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java (hunk 7 above)
Comment (git-tanuj) [reply extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java]: Sure, made the changes
Thank you

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/model/PolicyDefinitionDto.java (hunk 8 above)
Comment (git-tanuj) [reply extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/model/PolicyDefinitionDto.java]: Sure, made the changes.
Thank you

**Comment by paullatzelsperger** [comment]
@git-masoud and/or @git-tanuj before I take a deeper look a the PR, could I ask you to do two things:
1. resolve conversations if the requested change was made
2. re-request a review from the original reviewer. 

This keeps PRs neat and clean, and it puts the PR in reviewers' todo list.
Thanks!

File: spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java (hunk 1 above)
Comment (jimmarino) [reply spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java:77]: I don't understand why not passing the policy id explicitly and using the id from the policy definition is related to a service implementation depending on a controller.

File: spi/control-plane/policy-spi/src/testFixtures/java/org/eclipse/edc/connector/policy/spi/testfixtures/store/PolicyDefinitionStoreTestBase.java (hunk 3 above)
Comment (jimmarino) [reply spi/control-plane/policy-spi/src/testFixtures/java/org/eclipse/edc/connector/policy/spi/testfixtures/store/PolicyDefinitionStoreTestBase.java:102]: That behavior does not seem correct. Let's as @paullatzelsperger why that is the case.

File: spi/common/core-spi/src/main/java/org/eclipse/edc/spi/event/policydefinition/PolicyDefinitionUpdated.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+public class PolicyDefinitionUpdated extends PolicyDefinitionEvent<PolicyDefinitionUpdated.Payload> {
+
+    private PolicyDefinitionUpdated() {
+    }
+
+    /**
+     * This class contains all event specific attributes of a PolicyDefinition Update Event
+     *
+     */
+    public static class Payload extends PolicyDefinitionEvent.Payload {
```
[the remark this answers is not in the document]
Comment (jimmarino) [reply spi/common/core-spi/src/main/java/org/eclipse/edc/spi/event/policydefinition/PolicyDefinitionUpdated.java:29]: Please resolve this so it passes

[the remark this answers is not in the document]
Comment (jimmarino) [reply spi/common/core-spi/src/main/java/org/eclipse/edc/spi/event/policydefinition/PolicyDefinitionUpdated.java:29]: Please resolve this so it passes

**Comment by git-masoud** [comment]
> @git-masoud and/or @git-tanuj before I take a deeper look a the PR, could I ask you to do two things:
> 
>     1. resolve conversations if the requested change was made
> 
>     2. re-request a review from the original reviewer.
> 
> 
> This keeps PRs neat and clean, and it puts the PR in reviewers' todo list. Thanks!



> @git-masoud and/or @git-tanuj before I take a deeper look a the PR, could I ask you to do two things:
> 
>     1. resolve conversations if the requested change was made
> 
>     2. re-request a review from the original reviewer.
> 
> 
> This keeps PRs neat and clean, and it puts the PR in reviewers' todo list. Thanks!

The conversations have been resolved, but we still need your input on the remaining ones. As we used your approach as a reference in #PR 2156, we would greatly appreciate it if you could review them and provide feedback.

> @git-masoud and/or @git-tanuj before I take a deeper look a the PR, could I ask you to do two things:
> 
>     1. resolve conversations if the requested change was made
> 
>     2. re-request a review from the original reviewer.
> 
> 
> This keeps PRs neat and clean, and it puts the PR in reviewers' todo list. Thanks!

The conversations have been resolved, but we still need your input on the remaining ones. As we used your approach as a reference in #PR 2156, we would greatly appreciate it if you could review them and provide feedback.

**Comment by git-masoud** [comment]
Thanks @paullatzelsperger. The conversations have been resolved, but we still need your input on the remaining ones. As we used your approach as a reference in #PR [2156](https://github.com/eclipse-edc/Connector/pull/2516), we would greatly appreciate it if you could review them and provide feedback.

**Comment by paullatzelsperger** [comment]
> Thanks @paullatzelsperger. The conversations have been resolved, but we still need your input on the remaining ones. As we used your approach as a reference in #PR [2156](https://github.com/eclipse-edc/Connector/pull/2516), we would greatly appreciate it if you could review them and provide feedback.

I'll do that shortly, prob tomorrow or Friday

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
Regarding tests: I think the coverage can be improved a bit, e.g. adding, removing, replaceing policy fields (permissions, prohibitions...)

File: spi/common/core-spi/src/main/java/org/eclipse/edc/spi/event/policydefinition/PolicyDefinitionUpdated.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply spi/common/core-spi/src/main/java/org/eclipse/edc/spi/event/policydefinition/PolicyDefinitionUpdated.java:29]: we have the same issue on all the events, IMO we can just resolve this code scanning item

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java [hunk 10]
```diff
@@ -109,6 +112,23 @@ public IdResponseDto createPolicy(PolicyDefinitionRequestDto requestDto) {
 
     }
 
+    @PUT
+    @Path("{policyId}")
+    @Override
+    public void updatePolicy(@PathParam("policyId") String policyId, PolicyDefinitionUpdateDto updatedPolicyDefinition) {
```
[the remark this answers is not in the document]
Comment (git-masoud) [reply extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java]: Does it make sense to include @valid in the create function as well?

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/model/PolicyDefinitionUpdateDto.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+import org.eclipse.edc.policy.model.Policy;
+
+@JsonDeserialize(builder = PolicyDefinitionUpdateDto.Builder.class)
+public class PolicyDefinitionUpdateDto extends PolicyDefinitionDto {
+
+    private PolicyDefinitionUpdateDto() {
+    }
+
+    public Policy getPolicy() {
```
Comment (paullatzelsperger) [inline extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/model/PolicyDefinitionUpdateDto.java:28]: this getter should also be on the base class

File: extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply extensions/control-plane/api/management-api/policy-definition-api/src/main/java/org/eclipse/edc/connector/api/management/policy/PolicyDefinitionApiController.java]: yes, could you also add it there? thx

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
one minor thing, then I'm good with it.

File: spi/common/core-spi/src/main/java/org/eclipse/edc/spi/event/policydefinition/PolicyDefinitionUpdated.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply spi/common/core-spi/src/main/java/org/eclipse/edc/spi/event/policydefinition/PolicyDefinitionUpdated.java:29]: @git-masoud @git-tanuj We discussed internally the possibilities to refactor the `Event`/`Payload` stuff. for now, we'll just dismiss the alert.

File: spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java (hunk 1 above)
Comment (paullatzelsperger) [reply spi/control-plane/control-plane-spi/src/main/java/org/eclipse/edc/connector/spi/policydefinition/PolicyDefinitionService.java:77]: there is issue #2531 which takes care of that. can we resolve this then?

**Review by paullatzelsperger — APPROVED** [review]

File: spi/control-plane/policy-spi/src/testFixtures/java/org/eclipse/edc/connector/policy/spi/testfixtures/store/PolicyDefinitionStoreTestBase.java (hunk 3 above)
Comment (paullatzelsperger) [reply spi/control-plane/policy-spi/src/testFixtures/java/org/eclipse/edc/connector/policy/spi/testfixtures/store/PolicyDefinitionStoreTestBase.java:102]: similarly to the comment above, issue https://github.com/eclipse-edc/Connector/issues/2531 should take care of that

**Comment by paullatzelsperger** [comment]
@git-masoud @git-tanuj there are some merge conflicts, could you resolve them?

**Review by jimmarino — APPROVED** [review]

**Pull request merged**